lib/TPP/Conversion/ConvertVectorToXsmm/ConvertTranspose.cpp

```diff
@@ -0,0 +1,253 @@
+//===- ConvertTranspose.cpp ----------------------------------*- C++ -*-===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#include "TPP/Conversion/ConvertVectorToXsmm/ConvertTranspose.h"
+#include "TPP/Dialect/Xsmm/XsmmUtils.h"
+#include "TPP/Transforms/Utils/VNNIUtils.h"
+#include "TPP/Transforms/Utils/ValueUtils.h"
+#include "mlir/Dialect/Func/IR/FuncOps.h"
+#include "mlir/Dialect/LLVMIR/LLVMDialect.h"
+#include "mlir/Dialect/Linalg/IR/LinalgInterfaces.h"
+#include "mlir/Dialect/Linalg/Utils/Utils.h"
+#include "mlir/Dialect/MemRef/IR/MemRef.h"
+#include "mlir/Dialect/MemRef/Utils/MemRefUtils.h"
+#include "mlir/Dialect/PDL/IR/PDL.h"
+#include "mlir/Dialect/PDLInterp/IR/PDLInterp.h"
+#include "mlir/Dialect/Transform/PDLExtension/PDLExtensionOps.h"
+#include "mlir/Dialect/Vector/IR/VectorOps.h"
+#include "mlir/IR/BuiltinDialect.h"
+#include "mlir/IR/Operation.h"
+#include "mlir/IR/Value.h"
+#include "mlir/Interfaces/InferTypeOpInterface.h"
+#include "mlir/Interfaces/SideEffectInterfaces.h"
+#include "mlir/Parser/Parser.h"
+#include "mlir/Pass/Pass.h"
+#include "mlir/Transforms/GreedyPatternRewriteDriver.h"
+#include "llvm/ADT/TypeSwitch.h"
+#include "llvm/Support/Debug.h"
+using namespace mlir;
+using namespace mlir::vector;
+using namespace mlir::linalg;
+using namespace mlir::func;
+
+#include "ConvertTransposePDLLPatterns.h.inc"
+
+#define DEBUG_TYPE "convert-transpose"
+
+pair<Operation *, Operation *>
+getTransposeOperation(PatternRewriter &rewriter, xsmm::UnaryInfo &unaryInfo,
+                      Type outputType, xsmm::UnaryKind unaryKind,
+                      Operation *input, Operation *output,
+                      Operation *transposeOp) {
+  std::string dispatchName = "xsmm_unary_dispatch";
+  std::string invokeName = "xsmm_unary_invoke";
+  Location loc = transposeOp->getLoc();
+
+  ModuleOp module = transposeOp->getParentOfType<ModuleOp>();
+  auto dtype =
+      xsmm::utils::getDataType(rewriter, transposeOp->getOperand(0).getType());
+
+  auto functionOp = transposeOp->getParentOfType<func::FuncOp>();
+  SmallVector<Value, 10> dispatchOperands;
+  SmallVector<Type, 10> dispatchOperandTypes;
+  OpBuilder::InsertionGuard guard(rewriter);
+  rewriter.setInsertionPoint(&*functionOp.getBody().op_begin());
+  xsmm::UnaryKindAttr kind =
+      xsmm::UnaryKindAttr::get(rewriter.getContext(), unaryKind);
+  IntegerType integer64 = IntegerType::get(rewriter.getContext(), 64);
+  dispatchOperands.push_back(rewriter.create<arith::ConstantOp>(
+      loc, integer64, cast<TypedAttr>(kind)));
+  dispatchOperandTypes.push_back(integer64);
```

Comment:
Can't you really get the type from the value list?

Reply:
I'll change it to get the type from the value list.

Reply:
Done.